repos/system_upgrade/el8toel9/actors/opensslproviders/libraries/add_provider.py

```diff
@@ -0,0 +1,101 @@
+    """Add lines to the list of lines. Breaking possible newlines onto separate items"""
+    for l in add.split("\n"):
+        lines.append("{}\n".format(l))
+    return lines
+
+
+class NotFoundException(Exception):
+    pass
+
+
+def _replace(lines, search, replace, comment=None, backup=False, fail_on_error=True):
+    """Replace the lines (ignoring leading and trailing whitespace) matching `search` regex
+    in the given file (as `lines` collected using readlines()) with the `replace` line optional
+    comment is added on line preceeding the change. """
+    res = []
+    found = False
+    for line in lines:
+        if re.search(search, line.strip()):
+            if comment:
+                res.append(comment)
+            if backup:
+                res.append("# {}".format(line))
+            res = _add_lines(res, replace)
+            found = True
+        else:
+            res.append(line)
+    if not found and fail_on_error:
+        raise NotFoundException("The pattern {} not found.".format(search))
+    return res
+
+
+def _append(lines, add, comment=None):
+    if comment:
+        lines.append(comment)
+    return _add_lines(lines, add)
+
+
+def _modify_file(f, fail_on_error=True):
+    lines = f.readlines()
+    lines = _replace(lines, r"openssl_conf\s*=\s*default_modules",
+                     "openssl_conf = openssl_init",
+                     LEAPP_COMMENT, True, fail_on_error)
+    lines = _replace(lines, r"\[\s*default_modules\s*\]",
+                     "[openssl_init]\n"
+                     "providers = provider_sect",
+                     LEAPP_COMMENT, True, fail_on_error)
+    lines = _append(lines, APPEND_STRING, LEAPP_COMMENT)
+    f.seek(0)
+    f.write(''.join(lines))
+
+
+def process(openssl_messages):
+    """ The main logic
+        * read the file
+        * replace the required chunks
+        * write the file
+    """
+    config = next(openssl_messages, None)
+    if list(openssl_messages):
+        api.current_logger().warning('Unexpectedly received more than one OpenSslConfig message.')
+    if not config:
+        raise StopActorExecutionError(
+            'Could not check openssl configuration', details={'details': 'No OpenSslConfig facts found.'}
+        )
+
+    # If the configuration file was not modified, the rpm update will bring the new
+    # changes by itself -- do not change the file now.
+    if not config.modified:
+        return
+
+    # otherwise modify the file as announced in actors/opensslconfigcheck/actor.py
+    api.current_logger().debug('Modifying the {}.'.format(CONFIG))
+    try:
+        with open(CONFIG, 'r+') as f:
+            _modify_file(f)
+    except (OSError, IOError) as error:
+        api.current_logger().error('Failed to modify the file {}: {} '.format(CONFIG, error))
+    except NotFoundException as error:
+        raise StopActorExecutionError(
+            'Could not update {}'.format(CONFIG), details={'details': error}
+        )
```

Comment:
Discussed to exlucde StopActorExecutionError and instead:
- apply changes to a-file.leapp_backup
- if everything applied correctly, replace the original file
- on an error create the POST upgrade report (reporting.Groups.POST)
- log error
- on an error == OSError, IOError, NotFoundException, ...
-

Reply:
Changed to reflect the above. Actualy, the replace is done in memory and written only if all steps work so we do not need the backup.